### PR TITLE
Sidebar resizer all inclusive: A Demo

### PR DESCRIPTION
**Pull request by BurningTreeC**
Hello all,

based on the PR #8644 I made this little demo PR that has performance instrumentation enabled
I thought why should we refresh the stylesheets every time a `$:/state/` (other than the `$:/state/sidebar`) or `$:/temp` tiddler changes.
It also doesn't refresh the UI whenever the `$:/state/sidebar/dimensions` tiddler changes (and the `$:/state/sidebar/resizing` tiddler exists).

Have a look at the performance gains in comparison to #8644 , especially when there are like 20 tiddlers open for editing

:+1:

### CODE REVIEW

**Comment by pmario** [comment]
I do like the resizer. But I do not like the increase at the styelRefresh. 

If I am in tiddler edit-mode. Every keystroke styleRefresh needs 8-15ms if the PR version runs

The docs downloaded from TW-com need about 4-9ms -- I do not know where that change comes from.

The mainRefresh is about the same if I switch between **"right sidebar" -> More -> System** and **More -> Shadows**

**Comment by BurningTreeC** [comment]
Hi @pmario, now on my machine I'm at 5-8ms styleRefresh if you type in the editor
The added times come from the list and view widgets in the RootStylesheet I think, each widget refreshes...

**Comment by pmario** [comment]
There is some improvement. It's more constant now but still 8-12ms for styleRefres on my machine.

TW5-com is 5-8ms for styleRefresh. Obviously no changes here.

**Comment by BurningTreeC** [comment]
A question @pmario ... do you compare the current tw5-com **prerelease** with this PR? I've seen that times get worse with the amount of tiddlers in the wiki. Also, if you compare please compare either both single-file wikis or both nodejs versions.

**Comment by BurningTreeC** [comment]
Now the styleRefresh when editing tiddlers is **0**

File: core/ui/RootStylesheet.tid [hunk 1]
```diff
@@ -4,7 +4,7 @@ code-body: yes
 \import [subfilter{$:/core/config/GlobalImportFilter}]
 \whitespace trim
 \rules only filteredtranscludeinline transcludeinline html
-<$refresh-blocker refreshBlockerList="[all[tiddlers+shadows]is[draft]] [all[tiddlers]prefix[$:/state/]] [all[tiddlers]prefix[$:/temp/]] -[[$:/state/sidebar]] -[[$:/state/sidebar/dimensions]] -[[$:/state/sidebar/resizing]] -[[-$:/state/popup/menubar/hamburger]] +[is[unique]]">
+<$refresh-blocker refreshBlockerList="[all[tiddlers]is[draft]] [all[tiddlers]prefix[$:/state/]] [all[tiddlers]prefix[$:/temp/]] -[[$:/state/sidebar]] -[[$:/state/sidebar/dimensions]] -[[$:/state/sidebar/resizing]] -[[$:/state/popup/menubar/hamburger]]">
```
Comment (pmario) [inline core/ui/RootStylesheet.tid:7]: IMO `[is[draft]] [prefix[$:/state/]] [prefix[$:/temp/]] -[[$:/state/sidebar]] -[[$:/state/sidebar/dimensions]] -[[$:/state/sidebar/resizing]] -[[$:/state/popup/menubar/hamburger]]` should have the same effect

File: core/ui/RootStylesheet.tid [hunk 2]
```diff
@@ -4,12 +4,14 @@ code-body: yes
...
-<$let currentTiddler={{$:/language}} languageTitle={{!!name}}>
-	<$list filter="[all[shadows+tiddlers]!is[draft]tag[$:/tags/Stylesheet]]">
-		<$refresh-blocker refreshBlockerList={{{ [<currentTiddler>get[refresh-blocker-list]] }}}>
-			<style type="text/css">
-				<$view format={{{ [<currentTiddler>tag[$:/tags/Stylesheet/Static]then[text]else[plainwikified]] }}} mode="block"/>
-			</style>
-		</$refresh-blocker>
-	</$list>
-</$let>
+<$refresh-blocker refreshBlockerList="[all[tiddlers+shadows]is[draft]]">
```
Comment (pmario) [inline core/ui/RootStylesheet.tid]: Shadow tiddlers can never be drafts. A draft tiddler is always part of the wiki-store.

File: editions/tw5.com/tiddlers/_tw_shared/styles.tid [hunk 3]
```diff
@@ -1,5 +1,6 @@
 title: $:/_tw_shared/styles
 tags: $:/tags/Stylesheet TiddlyWikiSitesMenu
+refresh-blocker-list: [all[tiddlers]prefix[$:/state/]] [all[tiddlers]prefix[$:/temp/]] [all[tiddlers]is[draft]]
```
Comment (pmario) [inline editions/tw5.com/tiddlers/_tw_shared/styles.tid:3]: IMO filter can be optimized for all that fields

**Comment by pmario** [comment]
Timing looks good now.

**Comment by pmario** [comment]
Did you merge the latest master branch. There are new tests now. -- Did you check them yet?

File: core/ui/RootStylesheet.tid (hunk 1 above)
Comment (BurningTreeC) [reply core/ui/RootStylesheet.tid:7]: Yes thank you!

File: core/ui/RootStylesheet.tid (hunk 2 above)
Comment (BurningTreeC) [reply core/ui/RootStylesheet.tid]: Oh you're right :slightly_smiling_face:

File: editions/tw5.com/tiddlers/_tw_shared/styles.tid (hunk 3 above)
Comment (BurningTreeC) [reply editions/tw5.com/tiddlers/_tw_shared/styles.tid:3]: will do!

**Comment by BurningTreeC** [comment]
> Did you merge the latest master branch. There are new tests now. -- Did you check them yet?

Hi @pmario - yes I merged the latest master
The `./tiddlywiki.js editions/test --test` command fails - but not because of the view widget tests

**Comment by BurningTreeC** [comment]
The amount of files changed in this PR is because I added the `refresh-blocker-list` field to all stylesheets (also plugin stylesheets) and `refresh-blocker` widgets to all PageTemplates

**Comment by BurningTreeC** [comment]
By the way, using the `refresh-blocker` widget we can mitigate many performance issues.
It's true that it needs to be used with caution.
But I was for instance able to reduce the mainRefresh of the tag-picker with 100 edit tiddlers open in the story river to 120ms when going down/up the list. On tiddlywiki.com with 100 edit tiddlers open in the story river the tag-picker becomes nearly unusable.

**Comment by Jermolene** [comment]
Thanks @BurningTreeC I am pleased that you are pursuing this. A resizer for the sidebar is something that users would expect and it would be great to be able to offer it.

The approach you've taken to reduce the impact of the refresh cycle is interesting, and proves that the basic approach is sound. However, I don't like the special handling for specific state tiddlers. I think we might need to step back and try to optimise the refresh cycle more generally. Our goal should be to be able to hit 60 fps on modern machines. Then we can start to use these techniques to offer many more kinds of interactive experiences. Ultimately, every fixed number in our stylesheets (like the margins on tiddlers) should be interactively adjustable: tiddler padding and margins, the split between the editor and the preview etc.

In order to optimise anything one needs to be able to measure it, so the first thing we need is a test rig of some kind that lets us explore the impact of specific optimisation ideas. Perhaps we should run the tests in Puppeteer (like https://github.com/TiddlyWiki/tiddlywiki5-performance-test-rig/) to be closer to the end user experience, or perhaps it would be sufficient to run the performance rig entirely in Node.js. The test rig would work by loading up a set of test tiddlers, instantiating a render tree and then times the refresh when various different tiddlers are modified.

We need to understand exactly where all the time goes. I know that you have been experimenting along these lines, but I think there is great value in having a robust test rig that everyone can use.

I think there are still quite a few plausible generic optimisation ideas:

* During the refresh cycle we could detect subtrees that have completely static content, and do not need to be visited for refresh at all
* The filter mechanism already caches the compiled filter expressions, but there is currently no attempt to cache results of filter execution
* The refresh cycle could track

**Comment by BurningTreeC** [comment]
I'm closing this PR now

**Pull request closed without merging**